tailfs/fileserver.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package tailfs
+
+import (
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/tailscale/xnet/webdav"
+	"tailscale.com/tailfs/shared"
+)
+
+// FileServer is a standalone WebDAV server that dynamically serves up shares.
+// It's typically used in a separate process from the actual Tailfs server to
+// serve up files as an unprivileged user.
+type FileServer struct {
+	l             net.Listener
+	shareHandlers map[string]http.Handler
+	sharesMu      sync.RWMutex
+}
+
+// NewFileServer constructs a FileServer.
+//
+// The server attempts to listen at a random address on 127.0.0.1.
+// The listen address is available via the Addr() method.
+//
+// The server has to be told about shares before it can serve them. This is
+// accomplished either by calling SetShares(), or locking the shares with
+// LockShares(), clearing them with ClearSharesLocked(), adding them
+// individually with AddShareLocked(), and finally unlocking them with
+// UnlockShares().
+//
+// The server doesn't actually process requests until the Serve() method is
+// called.
+func NewFileServer() (*FileServer, error) {
+	// path := filepath.Join(os.TempDir(), fmt.Sprintf("%v.socket", uuid.New().String()))
+	// l, err := safesocket.Listen(path)
+	// if err != nil {
+	// TODO(oxtoacart): actually get safesocket working in more environments (MacOS Sandboxed, Windows, ???)
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return nil, err
+	}
+	// }
+	return &FileServer{
+		l:             l,
+		shareHandlers: make(map[string]http.Handler),
+	}, nil
+}
+
+// Addr returns the address at which this FileServer is listening.
+func (s *FileServer) Addr() string {
+	return s.l.Addr().String()
+}
+
+// Serve() starts serving files and blocks until it encounters a fatal error.
+func (s *FileServer) Serve() error {
+	return http.Serve(s.l, s)
+}
+
+// LockShares locks the map of shares in preparation for manipulating it.
```

Comment:
This API worries me a lot with explicit lock and unlock methods. We're requiring the caller to manage the lock explicitly, and it will be hard for tools to help us enforce that. I'm not sure I see why that's necessary—most of the methods either don't need locks (Addr, Serve) or already handle locking (SetShares).

Is there a reason I'm missing why ClearShares and AddShare cannot also handle locking themselves? 

If for some reason we can't have separate `ClearShares` and `AddShare` methods that do their own locking, I'd suggest we do something like:

```go
func (s *FileServer) WithLock(f func()) {
   s.sharesMu.Lock()
   defer s.sharesMu.Unlock()
   f()
}

// e.g.:
s.WithLock(s.AddShareLocked)
```

But really, even that isn't great because it doesn't prevent the caller from calling the method directly anyway.

Reply:
We should definitely document that a lot more specifically, then, because it's a very brittle API in isolation. We should make it clear this has to be used with care, and why we're doing this so that someone doesn't copy-paste it and try to reuse the pattern where it's not needed.